File: Minimum Total Distance Traveled.cpp
class Solution {
public:
    long long t[100][100][100];
    long long solve(vector<int>& robot, vector<vector<int>>& factory, int i, int j, int count) {
        if (i < 0) return 0;
        if (j < 0) return 1LL << 40;
        if (t[i][j][count] != -1) return t[i][j][count];
        long long skip = solve(robot, factory, i, j - 1, 0);
        long long take = (count < factory[j][1]) ? abs(robot[i] - factory[j][0]) + solve(robot, factory, i - 1, j, count + 1) : 1LL << 40;
        return t[i][j][count] = min(skip, take);
    }

    long long minimumTotalDistance(vector<int>& robot, vector<vector<int>>& factory) {
        int r = robot.size(), f = factory.size();
        memset(t, -1, sizeof(t));
        sort(robot.begin(), robot.end());
        sort(factory.begin(), factory.end());
        return solve(robot, factory, r - 1, f - 1, 0);
    }
};

File: Minimum Total Distance Traveled/Minimum Total Distance Traveled.java
// Solution for Minimum Total Distance Traveled in JAVA

class Solution {

    private long solve(int robotIndex,
                       int positionIndex,
                       List<Integer> robot,
                       List<Integer> factoryPositions,
                       long[][] dp) {

        if (robotIndex >= robot.size())
            return 0;

        if (positionIndex >= factoryPositions.size())
            return (long) 1e18;

        if (dp[robotIndex][positionIndex] != -1)
            return dp[robotIndex][positionIndex];

        // Assign
        long assign =
                Math.abs(robot.get(robotIndex) - factoryPositions.get(positionIndex)) +
                solve(robotIndex + 1, positionIndex + 1,
                      robot, factoryPositions, dp);

        // Skip
        long skip =
                solve(robotIndex, positionIndex + 1,
                      robot, factoryPositions, dp);

        return dp[robotIndex][positionIndex] = Math.min(assign, skip);
    }

    public long minimumTotalDistance(List<Integer> robot, int[][] factory) {

        // Sort robots
        Collections.sort(robot);

        // Sort factories by position
        Arrays.sort(factory, (a, b) -> a[0] - b[0]);

        // Expand factory capacities
        List<Integer> factoryPositions = new ArrayList<>();

        for (int[] f : factory) {
            int position = f[0];
            int capacity = f[1];

            for (int i = 0; i < capacity; i++) {
                factoryPositions.add(position);
            }
        }

        int n = robot.size();
        int m = factoryPositions.size();

        long[][] dp = new long[n + 1][m + 1];
        for (long[] row : dp)
            Arrays.fill(row, -1);

        return solve(0, 0, robot, factoryPositions, dp);
    }
}
